go/parquet/file/page_writer.go

```diff
@@ -0,0 +1,466 @@
+}
+
+func (pw *serializedPageWriter) updateEncryption(moduleType int8) error {
+	switch moduleType {
+	case encryption.ColumnMetaModule:
+		pw.metaEncryptor.UpdateAad(encryption.CreateModuleAad(pw.metaEncryptor.FileAad(), moduleType, pw.rgOrdinal, pw.columnOrdinal, -1))
+	case encryption.DataPageModule:
+		encryption.QuickUpdatePageAad(pw.dataPageAAD, pw.pageOrdinal)
+		pw.dataEncryptor.UpdateAad(string(pw.dataPageAAD))
+	case encryption.DataPageHeaderModule:
+		encryption.QuickUpdatePageAad(pw.dataPageHeaderAAD, pw.pageOrdinal)
+		pw.metaEncryptor.UpdateAad(string(pw.dataPageHeaderAAD))
+	case encryption.DictPageHeaderModule:
+		pw.metaEncryptor.UpdateAad(encryption.CreateModuleAad(pw.metaEncryptor.FileAad(), moduleType, pw.rgOrdinal, pw.columnOrdinal, -1))
+	case encryption.DictPageModule:
+		pw.dataEncryptor.UpdateAad(encryption.CreateModuleAad(pw.dataEncryptor.FileAad(), moduleType, pw.rgOrdinal, pw.columnOrdinal, -1))
+	}
+	return xerrors.New("unknown module type in updateencryption")
+}
+
+func (pw *serializedPageWriter) Close(hasDict, fallback bool) error {
+	if pw.metaEncryptor != nil {
+		pw.updateEncryption(encryption.ColumnMetaModule)
+	}
+
+	chunkInfo := metadata.ChunkMetaInfo{
+		NumValues:        pw.nvalues,
+		DictPageOffset:   pw.dictPageOffset,
+		IndexPageOffset:  -1,
+		DataPageOffset:   pw.dataPageOffset,
+		CompressedSize:   pw.totalCompressed,
+		UncompressedSize: pw.totalUncompressed,
+	}
+	encodingStats := metadata.EncodingStats{
+		DictEncodingStats: pw.dictEncodingStats,
+		DataEncodingStats: pw.dataEncodingStats,
+	}
+	pw.metaData.Finish(chunkInfo, hasDict, fallback, encodingStats, pw.metaEncryptor)
+	_, err := pw.metaData.WriteTo(pw.sink)
+	return err
+}
+
+func (pw *serializedPageWriter) Compress(buf *bytes.Buffer, src []byte) []byte {
+	maxCompressed := pw.compressor.CompressBound(int64(len(src)))
+	buf.Grow(int(maxCompressed))
+	return pw.compressor.EncodeLevel(buf.Bytes(), src, pw.compressLevel)
+}
+
+var dataPageV1HeaderPool = sync.Pool{
+	New: func() interface{} { return format.NewDataPageHeader() },
+}
+
+func (pw *serializedPageWriter) setDataPageHeader(pageHdr *format.PageHeader, page *DataPageV1) {
+	pageHdr.Type = format.PageType_DATA_PAGE
+	hdr := dataPageV1HeaderPool.Get().(*format.DataPageHeader)
+	hdr.NumValues = page.nvals
+	hdr.Encoding = page.encoding
+	hdr.DefinitionLevelEncoding = page.defLvlEncoding
+	hdr.RepetitionLevelEncoding = page.repLvlEncoding
+	hdr.Statistics = page.statistics.ToThrift()
+	pageHdr.DataPageHeader = hdr
+	pageHdr.DataPageHeaderV2 = nil
+	pageHdr.DictionaryPageHeader = nil
+}
+
+var dataPageV2HeaderPool = sync.Pool{
+	New: func() interface{} { return format.NewDataPageHeaderV2() },
+}
+
+func (pw *serializedPageWriter) setDataPageV2Header(pageHdr *format.PageHeader, page *DataPageV2) {
+	pageHdr.Type = format.PageType_DATA_PAGE_V2
+	hdr := dataPageV2HeaderPool.Get().(*format.DataPageHeaderV2)
+	hdr.NumValues = page.nvals
+	hdr.NumNulls = page.nulls
+	hdr.NumRows = page.nrows
```

Comment:
Can you check if https://issues.apache.org/jira/browse/PARQUET-2066 applies here?

Reply:
I guess it would be in the caller.

Reply:
Good catch, you're right. The `buildDataPageV2` function passes `numValues` for both `nvals` and `nrows`. I'll see about fixing that so it handles nested values properly to get the correct number of rows.

Reply:
Commit 2686ea26fb93c676825c3ba54e55b895202784c5 fixes the issue with num rows and adds a unit test for it.